File: library/src/main/java/com/sw926/imagefileselector/ImageFileSelector.java
package com.sw926.imagefileselector;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

@SuppressWarnings("unused")
public class ImageFileSelector {

    @Nullable
    private Callback mCallback = null;
    private ImagePickHelper mImagePickHelper;
    private ImageCaptureHelper mImageCaptureHelper;
    private ImageCompressHelper mImageCompressHelper;

    public ImageCompressHelper.CompressParams compressParams;

    public ImageFileSelector(Context context) {

        String defaultOutputPath = context.getExternalCacheDir() + "/images/";
        compressParams = new ImageCompressHelper.CompressParams();
        compressParams.outputPath = defaultOutputPath;

        mImageCompressHelper = new ImageCompressHelper();
        mImageCompressHelper.setCallback(new ImageCompressHelper.Callback() {
            @Override
            public void onError() {
                if (mCallback != null) {
                    mCallback.onError(ErrorResult.error);
                }
            }

            @Override
            public void onSuccess(String file) {
                if (mCallback != null) {
                    mCallback.onSuccess(file);
                }

            }
        });

        Callback callback = new Callback() {
            @Override
            public void onError(ErrorResult errorResult) {
                if (mCallback != null) {

                    mCallback.onError(errorResult);
                }
            }

            @Override
            public void onSuccess(String file) {
                ImageCompressHelper.CompressJop jop = new ImageCompressHelper.CompressJop();
                jop.params = compressParams;
                jop.inputFile = file;
                mImageCompressHelper.compress(jop);
            }
        };

        mImagePickHelper = new ImagePickHelper(context);
        mImagePickHelper.setCallback(callback);

        mImageCaptureHelper = new ImageCaptureHelper();
        mImageCaptureHelper.setCallback(callback);
    }


    public void setOutPutPath(String outPutPath) {
        compressParams.outputPath = outPutPath;
    }

    public void setSelectFileType(String type) {
        mImagePickHelper.setType(type);
    }

    /**
     * 设置压缩后的文件大小
     *
     * @param maxWidth  压缩后文件宽度
     *                  *
     * @param maxHeight 压缩后文件高度
     */
    @SuppressWarnings("unused")
    public void setOutPutImageSize(int maxWidth, int maxHeight) {
        compressParams.maxWidth = maxWidth;
        compressParams.maxHeight = maxHeight;
    }

    /**
     * 设置压缩后保存图片的质量
     *
     * @param quality 图片质量 0 - 100
     */
    @SuppressWarnings("unused")
    public void setQuality(int quality) {
        compressParams.saveQuality = quality;
    }

    /**
     * set image compress format
     *
     * @param compressFormat compress format
     */
    @SuppressWarnings("unused")
    public void setCompressFormat(Bitmap.CompressFormat compressFormat) {
        compressParams.compressFormat = compressFormat;
    }

    public void onActivityResult(Context context, int requestCode, int resultCode, Intent data) {
        mImagePickHelper.onActivityResult(requestCode, resultCode, data);
        mImageCaptureHelper.onActivityResult(context, requestCode, resultCode, data);
    }

    public void onRequestPermissionsResult(Context context, int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == mImagePickHelper.getRequestCode()) {
            mImagePickHelper.onRequestPermissionsResult(requestCode, permissions, grantResults);
        } else if (requestCode == mImageCaptureHelper.getRequestCode()) {
            mImageCaptureHelper.onRequestPermissionsResult(context, requestCode, permissions, grantResults);
        }
    }

    public void onSaveInstanceState(Bundle outState) {
        if (mImageCaptureHelper != null) {
            mImageCaptureHelper.onSaveInstanceState(outState);
        }
        if (mImagePickHelper != null) {
            mImagePickHelper.onSaveInstanceState(outState);
        }
    }

    public void onRestoreInstanceState(Bundle outState) {
        if (mImageCaptureHelper != null) {
            mImageCaptureHelper.onRestoreInstanceState(outState);
        }
        if (mImagePickHelper != null) {
            mImagePickHelper.onRestoreInstanceState(outState);
        }
    }

    public void setCallback(@Nullable Callback callback) {
        mCallback = callback;
    }

    public void selectImage(Activity activity, int requestCode) {
        mImagePickHelper.selectorImage(activity, requestCode);
    }

    public void selectImage(Fragment fragment, int requestCode) {
        mImagePickHelper.selectImage(fragment, requestCode);
    }

    public void takePhoto(Activity activity, int requestCode) {
        mImageCaptureHelper.captureImage(activity, requestCode);
    }

    public void takePhoto(Fragment fragment, int requestCode) {
        mImageCaptureHelper.captureImage(fragment, requestCode);
    }


    public static void setDebug(Boolean debug) {
        AppLogger.DEBUG = debug;
    }

    /**
     * Created by sunwei on 10/11/2016 2:36 PM.
     */

    public interface Callback {
        void onError(ErrorResult errorResult);

        void onSuccess(String file);
    }
}

File: library/src/main/java/com/sw926/imagefileselector/ImageFileSelector.kt
@file:JvmName("ImageFileSelector")

package com.sw926.imagefileselector

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import java.lang.ref.WeakReference

class ImageFileSelector {

    private var resultListener: ImageFileResultListener? = null
    private val mImagePickHelper: ImagePickHelper
    private val mImageCaptureHelper: ImageCaptureHelper

    private var compressParams: CompressParams
    private val context: WeakReference<Context>

    constructor(activity: AppCompatActivity) {
        mImageCaptureHelper = ImageCaptureHelper(activity)
        mImagePickHelper = ImagePickHelper(activity)

        compressParams = CompressParams()
        context = WeakReference(activity)

        init()
    }

    constructor(fragment: Fragment) {
        mImageCaptureHelper = ImageCaptureHelper(fragment)
        mImagePickHelper = ImagePickHelper(fragment)

        compressParams = CompressParams()
        context = WeakReference(fragment.context)
        init()
    }

    private fun init() {
        val listener = object : ImageUriResultListener {
            override fun onSuccess(uri: Uri) {

                val context = this@ImageFileSelector.context.get()
                if (context == null) {
                    resultListener?.onError()
                    return
                }
                ImageUtils.compress(context, uri, compressParams) {
                    if (it != null) {
                        resultListener?.onSuccess(it)
                    } else {
                        resultListener?.onError()
                    }
                }
            }

            override fun onCancel() {
            }

            override fun onError() {
            }
        }

        mImagePickHelper.setListener(listener)
        mImageCaptureHelper.setListener(listener)
    }

    @Suppress("unused")
    fun setSelectFileType(type: String) {
        mImagePickHelper.setType(type)
    }

    /**
     * 设置压缩后的文件大小
     *
     * @param maxWidth  压缩后文件宽度
     * *
     * @param maxHeight 压缩后文件高度
     */
    fun setOutPutImageSize(maxWidth: Int, maxHeight: Int) {
        compressParams = compressParams.copy(maxWidth = maxWidth, maxHeight = maxHeight)
    }

    /**
     * 设置压缩后保存图片的质量
     *
     * @param quality 图片质量 0 - 100
     */
    @Suppress("unused")
    fun setQuality(quality: Int) {
        compressParams = compressParams.copy(saveQuality = quality)
    }

    /**
     * set image compress format
     *
     * @param compressFormat compress format
     */
    @Suppress("unused")
    fun setCompressFormat(compressFormat: Bitmap.CompressFormat) {
        compressParams = compressParams.copy(compressFormat = compressFormat)
    }


    fun setListener(listenerUri: ImageFileResultListener?) {
        resultListener = listenerUri
    }

    fun selectImage() {
        mImagePickHelper.pickImage()
    }

    fun takePhoto() {
        mImageCaptureHelper.takePicture()
    }

}
